src/Streamly/Internal/Data/Unfold/Resume.hs

```diff
@@ -0,0 +1,231 @@
+
+-- | Convert a StreamD stream into an Unfold.
+--
+-- /Internal/
+{-# INLINE_NORMAL fromStreamD #-}
+fromStreamD :: Monad m => Unfold m (Stream m a) a
+fromStreamD = Unfold step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step (UnStream step1 state1) = do
+        r <- step1 defState state1
+        return $ case r of
+            Yield x s -> Yield x (Stream step1 s)
+            Skip s    -> Skip (Stream step1 s)
+            Stop      -> Stop
+
+-------------------------------------------------------------------------------
+-- Parsing
+-------------------------------------------------------------------------------
+
+-- GHC parser does not accept {-# ANN type [] NoSpecConstr #-}, so we need
+-- to make a newtype.
+{-# ANN type List NoSpecConstr #-}
+newtype List a = List {getList :: [a]}
+
+{-# INLINE_NORMAL parseD #-}
+parseD
+    :: MonadThrow m
+    => ParserD.Parser m a b
+    -> Unfold m (Source s a) a
+    -> Source s a
+    -> m (b, Source s a)
+parseD
+    (ParserD.Parser pstep initial extract)
+    (Unfold ustep uinject uextract)
+    seed = do
+
+    state <- uinject seed
+    initial >>= go SPEC state (List [])
+
+    where
+
+    -- XXX currently we are using a dumb list based approach for backtracking
+    -- buffer. This can be replaced by a sliding/ring buffer using Data.Array.
+    -- That will allow us more efficient random back and forth movement.
+    {-# INLINE go #-}
+    go !_ st buf !pst = do
+        r <- ustep st
+        case r of
+            Yield x s -> do
+                pRes <- pstep pst x
+                case pRes of
+                    Partial 0 pst1 -> go SPEC s (List []) pst1
+                    Partial n pst1 -> do
+                        assert (n <= length (x:getList buf)) (return ())
+                        let src0 = Prelude.take n (x:getList buf)
+                            src  = Prelude.reverse src0
+                        gobuf SPEC s (List []) (List src) pst1
+                    Continue 0 pst1 -> go SPEC s (List (x:getList buf)) pst1
+                    Continue n pst1 -> do
+                        assert (n <= length (x:getList buf)) (return ())
+                        let (src0, buf1) = splitAt n (x:getList buf)
+                            src  = Prelude.reverse src0
+                        gobuf SPEC s (List buf1) (List src) pst1
+                    Done n b -> do
+                        assert (n <= length (x:getList buf)) (return ())
+                        let src0 = Prelude.take n (x:getList buf)
+                            src  = Prelude.reverse src0
+                        s1 <- uextract s
+                        return (b, Source.unread src s1)
+                    Error err -> throwM $ ParseError err
+            Skip s -> go SPEC s buf pst
+            Stop   -> do
+                b <- extract pst
+                -- XXX we should return the remaining buffer
+                return (b, Source.source Nothing)
```

Comment:
Why don't you?

Reply:
Also, I'm only looking at `Done` in `parseD`. I'm assuming everything else is the same.

Reply:
It was before we had a Maybe in the source, now we can return the buffer safely I guess. Done.